specs/request-handler-plugins.md

```diff
@@ -0,0 +1,116 @@
+  ],
+  "msGraphProxy": {
+    // config options specific to the msGraphProxy
+  }
+```
+
+Handlers will executed in the order that they are listed in the handlers configuration array. At least one handlers must be present in the configuration, otherwise the proxy will exit with an error informing the user that their configuration is invalid and contains no handlers.
+
+## Handlers
+
+Handler classes must inherit from the provider abstract base class `ProxyHandler`, proposed implementation listed below for reference, and must provide an implementation for the `ExecuteInternal` method.
+Handlers can set the response which will be provided by the proxy using the `GenericResponse` method of the provided `SessionEventArgs` object, when a handler sets the response it should also set the `IsComplete` property of the provided `ResponseState` to prevent subsequently executed handler from modifying the response.
+
+```cs
+using System.CommandLine;
+using System.CommandLine.Invocation;
+using System.Text.RegularExpressions;
+using Microsoft.Extensions.Configuration;
+using Titanium.Web.Proxy.EventArguments;
+
+public abstract class ProxyHandler
+{
+    private readonly IEnumerable<Regex> _urlsToWatch;
+    private readonly IConfigurationSection _config;
+
+    public ProxyHandler(IEnumerable<Regex> urlsToWatch, IConfigurationSection config)
+    {
+        this._urlsToWatch = urlsToWatch ?? throw new ArgumentNullException(nameof(urlsToWatch));
+        if (!urlsToWatch.Any())
+            throw new ArgumentException($"{nameof(urlsToWatch)} cannot be empty", nameof(urlsToWatch));
+        this._config = config;
+    }
+
+    /**
+     * Use this method to add any command line options to the proxy
+     */
+    public abstract void AddCommandLineOptions(RootCommand command);
+
+    /**
+     * Read options from command line args and update internal configuration
+     */
+    public abstract int UpdateConfiguration(InvocationContext context);
+
+    /**
+     * Use this method to provide a concrete implmentation for a request handler.
+     * Set ResponseState.HasBeenSet to prevent additional handlers from modifying the response
+     */
+    protected abstract void ExecuteInternal(SessionEventArgs e, ResponseState state);
+
+    /**
+     * Default implementation to determine if the handler should be triggered.
+     * Override if you need to ignore the ResponseState.HasBeenSet, typically this would be done
+     * when your handler does not mutate the response and must be executed, ususally this can be
+     * better handled by configuring the handler to run earlier in the set of configured handlers
+     */
+    protected virtual bool ShouldTrigger(SessionEventArgs e, ResponseState state) =>
+        !state.HasBeenSet &&
+        this._urlsToWatch.Any(u => u.IsMatch(e.HttpClient.Request.RequestUriString));
+
+    /**
+     * Public entrypoint called by the Proxy for each request.
+     */
+    public void Excecute(SessionEventArgs e, ResponseState state)
```

Comment:
Shall we also allow plugins to process the response of a request that we passed through to Graph? Right now, we have the limitation that we [stop our processing pipe](https://github.com/microsoftgraph/msgraph-developer-proxy/blob/e8ca93943a8ab101cff7dbc1d9ad50c7cd272345/msgraph-developer-proxy/ChaosEngine.cs#L311-L315) when we decided that the request should be passed to Graph. If that request got throttled, our SDK guidance wouldn't run. If we're to allow for this, we'd need to consider using an event-based handling rather than a single method.

Reply:
In principle I like the idea of allowing a plugin to tap into the response being sent, be it from the underlying API or as a result of intercepting the request and setting a response in a plugin.

I'm just not sure what the use case for this would be.

Reply:
Resolved via change to event-based model